### PR TITLE
GH-215: Add timeout to health indicator

### PR DESCRIPTION
**Pull request by garyrussell**
Resolves https://github.com/spring-cloud/spring-cloud-stream-binder-kafka/issues/215

### CODE REVIEW

File: spring-cloud-stream-binder-kafka/src/main/java/org/springframework/cloud/stream/binder/kafka/KafkaBinderHealthIndicator.java [hunk 1]
```diff
@@ -33,42 +38,89 @@
...
-				List<PartitionInfo> partitionInfos = metadataConsumer.partitionsFor(topic);
-				for (PartitionInfo partitionInfo : partitionInfos) {
-					if (this.binder.getTopicsInUse().get(topic).getPartitionInfos().contains(partitionInfo)
-							&& partitionInfo.leader()
-									.id() == -1) {
-						downMessages.add(partitionInfo.toString());
+		final CountDownLatch latch = new CountDownLatch(1);
+		final AtomicReference<Health> health = new AtomicReference<>();
+		ExecutorService exec = Executors.newSingleThreadExecutor();
+		exec.execute(new Runnable() {
```
Comment (artembilan) [inline spring-cloud-stream-binder-kafka/src/main/java/org/springframework/cloud/stream/binder/kafka/KafkaBinderHealthIndicator.java]: I believe that it works as you designed, but wouldn't `ExecutorService.submit()` be cleaner solution?
At least we won't need extra `CountDownLatch` variable.

File: spring-cloud-stream-binder-kafka/src/main/java/org/springframework/cloud/stream/binder/kafka/KafkaBinderHealthIndicator.java [hunk 2]
```diff
@@ -33,42 +40,87 @@
...
-		try (Consumer<?, ?> metadataConsumer = consumerFactory.createConsumer()) {
-			Set<String> downMessages = new HashSet<>();
-			for (String topic : this.binder.getTopicsInUse().keySet()) {
-				List<PartitionInfo> partitionInfos = metadataConsumer.partitionsFor(topic);
-				for (PartitionInfo partitionInfo : partitionInfos) {
-					if (this.binder.getTopicsInUse().get(topic).getPartitionInfos().contains(partitionInfo)
-							&& partitionInfo.leader()
-									.id() == -1) {
-						downMessages.add(partitionInfo.toString());
+		final AtomicReference<Health> health = new AtomicReference<>();
```
Comment (artembilan) [inline spring-cloud-stream-binder-kafka/src/main/java/org/springframework/cloud/stream/binder/kafka/KafkaBinderHealthIndicator.java]: But we don't need `AtomicReference` as well.
We can use `submit(Callable<T>)`  and return `Health` from there and `get()` it from the `Future`, therefore.

Comment (garyrussell) [reply spring-cloud-stream-binder-kafka/src/main/java/org/springframework/cloud/stream/binder/kafka/KafkaBinderHealthIndicator.java]: Doh

**Comment by artembilan** [comment]
Cherry-picked to `1.2.x` as https://github.com/spring-cloud/spring-cloud-stream-binder-kafka/commit/657e8a2d7c54866250b7ee716b217705c0a61b9a

**Pull request merged**